vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,195 @@
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import copy
+import enum
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
+
+import torch
+import xgrammar as xgr
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.transformers_utils.tokenizer_group import init_tokenizer_from_configs
+
+if TYPE_CHECKING:
+    from vllm.v1.request import Request
+
+logger = init_logger(__name__)
+
+
+class GuidedDecodingOptions(enum.Enum):
+    json = enum.auto()
+    json_object = enum.auto()
+    regex = enum.auto()
+    grammar = enum.auto()
+    choice = enum.auto()
+
+
+GuidedDecodingKey = Tuple[GuidedDecodingOptions, str]
+MAX_ROLLBACK_TOKENS = 100
+
+
+def apply_bitmask(logits: torch.Tensor, vocab_mask: torch.Tensor,
+                  indices: List[int]) -> None:
+    xgr.apply_token_bitmask_inplace(logits, vocab_mask, indices=indices)
+
+
+@dataclass(slots=True, unsafe_hash=True)
+class Grammar:
+    # NOTE: This would be a generic-enough class for
+    # supporting different backends, in the future.
+    # For now, just xgrammar.
+    #
+    # TODO: support max_rollback_tokens
+    # https://xgrammar.mlc.ai/docs/api/python/index.html#xgrammar.GrammarMatcher.find_jump_forward_string
+    # for jump-forward decoding
+
+    vocab_size: int
+    matcher: xgr.GrammarMatcher = field(hash=False)
+    ctx: xgr.CompiledGrammar = field(hash=False)
+    max_rollback_tokens: int = field(default=MAX_ROLLBACK_TOKENS, kw_only=True)
+    num_processed_tokens: int = field(
+        default_factory=lambda: 0,
+        repr=False,
+        hash=False,
+        init=False,
+    )
+    _matcher_lock: threading.Lock = field(
+        default_factory=lambda: threading.Lock(),
+        repr=False,
+        init=False,
+        hash=False,
+    )
```

Comment:
I think this lock can be removed. We are not sharing a matcher between requests anymore (we can't), so the lock should not be necessary